### PR TITLE
add rows_per_output_partition

### PR DESCRIPTION
**Pull request by wearpants**
This adds a model-level output to control partition size.

### CODE REVIEW

File: tests/osp_pipeline/corpus/jobs/test_extract_text.py [hunk 1]
```diff
@@ -27,7 +27,7 @@ def texts(sc, spark, results):
     ])
 
     # Run job.
-    extract_text.PARTITIONS = 1
+    DocumentText.rows_per_output_partition = None
```
Comment (jodizzle) [inline tests/osp_pipeline/corpus/jobs/test_extract_text.py]: Is changing the partitioning in the tests important?  Tests seem to pass with or without this line.

File: osp_pipeline/core/models.py [hunk 2]
```diff
@@ -51,8 +54,26 @@ def build_df(cls, rdd):
     @classmethod
     def save_df(cls, df):
         """save a dataframe to the appropriate S3 path from `config`"""
+        log = logging.getLogger('osp_pipeline')
+        df = df.cache()
```
Comment (jodizzle) [inline osp_pipeline/core/models.py]: Do you need to assign here?  I thought `df.cache()` was just called as an expression by itself.

**Comment by jodizzle** [comment]
Is there any reason to not just search and replace all usages of `read` and `write` and fully remove those methods?

**Comment by wearpants** [comment]
On Feb 21, 2018 3:07 AM, "jodizzle" <notifications@github.com> wrote:

*@jodizzle* commented on this pull request.
------------------------------

In tests/osp_pipeline/corpus/jobs/test_extract_text.py
<https://github.com/opensyllabus/osp-pipeline/pull/148#discussion_r169558884>
:

> @@ -27,7 +27,7 @@ def texts(sc, spark, results):
     ])

     # Run job.
-    extract_text.PARTITIONS = 1
+    DocumentText.rows_per_output_partition = None

Is changing the partitioning in the tests important? Tests seem to pass
with or without this line.


Yes

**Comment by wearpants** [comment]
They are horked in tests

**Comment by jodizzle** [comment]
@wearpants Why is manually setting the partitioning in the extract text job important?  Does it need to be set in the tests for merge documents, find syllabi, etc.?

**Comment by wearpants** [comment]
No

**Comment by wearpants** [comment]
Hmm maybe

On Feb 21, 2018 1:47 PM, "Pete Fein" <pete@syste.ma> wrote:

No

**Comment by wearpants** [comment]
setting # of output partitions is only necessary in tests that write to disk (with a call to `save_df`), which we shouldn't be doing anyway.

This whole saving/testing interaction is a bit of a mess, and we don't have time to attend to it right now

**Comment by jodizzle** [comment]
I see, that makes some sense.

Just to clarify though: is the number of partitions relevant to the performance of the test?  Or is it just about making it deterministic?  Following the logic of `save_df` and comparing it to what's currently tested, it seems like the number of partitions would be 1 in both cases.

**Pull request merged**